File: frontend/src/apis/site.ts
import {request} from "../utils/requests.ts";

export interface SiteCapabilities {
    supports_ranking: boolean;
    supports_actor: boolean;
    supports_login: boolean;
    supports_downloads: boolean;
    supports_previews: boolean;
    supports_comments: boolean;
}

export interface SiteItem {
    id: number;
    spider_key: 'javdb' | 'javbus' | 'jav321' | 'dmm';
    name: string;
    priority: number;
    alternate_host?: string;
    status?: boolean;
    cookies?: string;
    capabilities: SiteCapabilities;
}

export interface SiteUpdate {
    id: number;
    priority: number;
    alternate_host?: string;
    status?: boolean;
    cookies?: string;
}

export async function getSites(): Promise<SiteItem[]> {
    const response = await request.get('/site/');
    return response.data.data;
}

export function modifySite(site: SiteUpdate) {
    return request.put('/site/', site)
}

export function testingSits() {
    return request.post('/site/testing');
}

export function getLoginPage(siteId: number) {
    return request.get(`/site/${siteId}/login/page`);
}

export function submitLogin(siteId: number, data: {
    cookies: string;
    authenticity_token: string;
    username: string;
    password: string;
    captcha: string;
}) {
    return request.post(`/site/${siteId}/login/submit`, data);
}

File: app/schema/site.py
from enum import StrEnum
from typing import Optional

from pydantic import BaseModel


class SpiderKey(StrEnum):
    JAVDB = 'javdb'
    JAVBUS = 'javbus'
    JAV321 = 'jav321'
    DMM = 'dmm'


class MetadataPriorityFieldKey(StrEnum):
    COVER = 'cover'
    RATING = 'rating'
    ACTORS = 'actors'


class SiteCapabilities(BaseModel):
    supports_ranking: bool
    supports_actor: bool
    supports_login: bool
    supports_downloads: bool
    supports_previews: bool
    supports_comments: bool


class Site(BaseModel):
    id: int
    spider_key: SpiderKey
    name: str
    priority: int
    alternate_host: Optional[str]
    status: Optional[bool]
    cookies: Optional[str]
    capabilities: SiteCapabilities


class SiteUpdate(BaseModel):
    id: int
    priority: int
    alternate_host: Optional[str]
    status: Optional[bool]
    cookies: Optional[str]


class LoginSubmit(BaseModel):
    cookies: str
    authenticity_token: str
    username: str
    password: str
    captcha: str


class MetadataPriorityFieldConfig(BaseModel):
    sites: list[SpiderKey]
    is_default: bool


class MetadataPriorityFieldsConfig(BaseModel):
    cover: MetadataPriorityFieldConfig
    rating: MetadataPriorityFieldConfig
    actors: MetadataPriorityFieldConfig


class MetadataPrioritySettings(BaseModel):
    global_sites: list[SpiderKey]
    fields: MetadataPriorityFieldsConfig


class MetadataPriorityFieldsUpdate(BaseModel):
    cover: list[SpiderKey]
    rating: list[SpiderKey]
    actors: list[SpiderKey]


class MetadataPriorityUpdate(BaseModel):
    fields: MetadataPriorityFieldsUpdate
